In [1]:
import pandas as pd
from decimal import Decimal

from carbon import CarbonSimulatorUI, __version__, __date__
print(f"Carbon Version v{__version__} ({__date__})", )

Carbon Version v2.2-BETA2 (03/Jan/2022)


# Carbon Simulation - Demo 5-2

In this demo we investigate the **Fast Router performance** against implemented code

## Init

Initialize a fast simulator

In [2]:
Sim = CarbonSimulatorUI(pair="ETH/USDC", verbose=False, raiseonerror=True)
FastSim = CarbonSimulatorUI(pair="ETH/USDC", verbose=False, matching_method='fast', raiseonerror=True)

Here we restrain the orders to just the top ones such that we can test the exact algo alongside

In [3]:
test_orders = pd.read_csv('orders.csv', dtype=str)
test_orders = test_orders[test_orders.id.isin(['0','1','2','3','251','250','6','248','247','246','10'])].copy()
test_orders

id  liquidity lowest_rate highest_rate current_rate
0      0  254814732         256          257          257
1      1  253827078         255          256          256
2      2  252839424         254          255          255
3      3  251851770         253          254          254
6      6  248888808         250          251          251
10    10  244938192         246          247          247
246  246  245925846         247          248          248
247  247  246913500         248          249          249
248  248  247901154         249          250          250
250  250  249876462         251          252          252
251  251  250864116         252          253          253

In [4]:
for i in test_orders.index:
    Sim.add_strategy('USDC', Decimal(test_orders.liquidity[i]), Decimal(test_orders.highest_rate[i]), Decimal(test_orders.lowest_rate[i]),  None, None, None)
    FastSim.add_strategy('USDC', Decimal(test_orders.liquidity[i]), Decimal(test_orders.highest_rate[i]), Decimal(test_orders.lowest_rate[i]),  None, None, None)

FastSim.state()['orders']

id     pair   tkn        y_int            y y_unit  disabled p_start  \
0    0  ETHUSDC  USDC  254814732.0  254814732.0   USDC     False   257.0   
1    1  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
2    2  ETHUSDC  USDC  253827078.0  253827078.0   USDC     False   256.0   
3    3  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
4    4  ETHUSDC  USDC  252839424.0  252839424.0   USDC     False   255.0   
5    5  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
6    6  ETHUSDC  USDC  251851770.0  251851770.0   USDC     False   254.0   
7    7  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
8    8  ETHUSDC  USDC  248888808.0  248888808.0   USDC     False   251.0   
9    9  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
10  10  ETHUSDC  USDC  244938192.0  244938192.0   USDC     False   247.0   
11  11  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
12  12  ETHUSDC  USDC  245925846.0  245925846.0   USDC     False   248.0   
13  13  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
14  14  ETHUSDC  USDC  246913500.0  246913500.0   USDC     False   249.0   
15  15  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
16  16  ETHUSDC  USDC  247901154.0  247901154.0   USDC     False   250.0   
17  17  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
18  18  ETHUSDC  USDC  249876462.0  249876462.0   USDC     False   252.0   
19  19  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
20  20  ETHUSDC  USDC  250864116.0  250864116.0   USDC     False   253.0   
21  21  ETHUSDC   ETH          0.0          0.0    ETH      True    None   

    p_end p_marg        p_unit  lid  
0   256.0  257.0  USDC per ETH    1  
1    None   None  USDC per ETH    0  
2   255.0  256.0  USDC per ETH    3  
3    None   None  USDC per ETH    2  
4   254.0  255.0  USDC per ETH    5  
5    None   None  USDC per ETH    4  
6   253.0  254.0  USDC per ETH    7  
7    None   None  USDC per ETH    6  
8   250.0  251.0  USDC per ETH    9  
9    None   None  USDC per ETH    8  
10  246.0  247.0  USDC per ETH   11  
11   None   None  USDC per ETH   10  
12  247.0  248.0  USDC per ETH   13  
13   None   None  USDC per ETH   12  
14  248.0  249.0  USDC per ETH   15  
15   None   None  USDC per ETH   14  
16  249.0  250.0  USDC per ETH   17  
17   None   None  USDC per ETH   16  
18  251.0  252.0  USDC per ETH   19  
19   None   None  USDC per ETH   18  
20  252.0  253.0  USDC per ETH   21  
21   None   None  USDC per ETH   20

In [5]:
FastSim.state()['orders']

id     pair   tkn        y_int            y y_unit  disabled p_start  \
0    0  ETHUSDC  USDC  254814732.0  254814732.0   USDC     False   257.0   
1    1  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
2    2  ETHUSDC  USDC  253827078.0  253827078.0   USDC     False   256.0   
3    3  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
4    4  ETHUSDC  USDC  252839424.0  252839424.0   USDC     False   255.0   
5    5  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
6    6  ETHUSDC  USDC  251851770.0  251851770.0   USDC     False   254.0   
7    7  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
8    8  ETHUSDC  USDC  248888808.0  248888808.0   USDC     False   251.0   
9    9  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
10  10  ETHUSDC  USDC  244938192.0  244938192.0   USDC     False   247.0   
11  11  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
12  12  ETHUSDC  USDC  245925846.0  245925846.0   USDC     False   248.0   
13  13  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
14  14  ETHUSDC  USDC  246913500.0  246913500.0   USDC     False   249.0   
15  15  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
16  16  ETHUSDC  USDC  247901154.0  247901154.0   USDC     False   250.0   
17  17  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
18  18  ETHUSDC  USDC  249876462.0  249876462.0   USDC     False   252.0   
19  19  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
20  20  ETHUSDC  USDC  250864116.0  250864116.0   USDC     False   253.0   
21  21  ETHUSDC   ETH          0.0          0.0    ETH      True    None   

    p_end p_marg        p_unit  lid  
0   256.0  257.0  USDC per ETH    1  
1    None   None  USDC per ETH    0  
2   255.0  256.0  USDC per ETH    3  
3    None   None  USDC per ETH    2  
4   254.0  255.0  USDC per ETH    5  
5    None   None  USDC per ETH    4  
6   253.0  254.0  USDC per ETH    7  
7    None   None  USDC per ETH    6  
8   250.0  251.0  USDC per ETH    9  
9    None   None  USDC per ETH    8  
10  246.0  247.0  USDC per ETH   11  
11   None   None  USDC per ETH   10  
12  247.0  248.0  USDC per ETH   13  
13   None   None  USDC per ETH   12  
14  248.0  249.0  USDC per ETH   15  
15   None   None  USDC per ETH   14  
16  249.0  250.0  USDC per ETH   17  
17   None   None  USDC per ETH   16  
18  251.0  252.0  USDC per ETH   19  
19   None   None  USDC per ETH   18  
20  252.0  253.0  USDC per ETH   21  
21   None   None  USDC per ETH   20

### Route by Source

#### AMM buys ETH

Lets do some simple checks against the exact algo and then verify against the fast router results

1. Trade an amount against exact

In [6]:
results = Sim.amm_buys('ETH',10000000, execute=False)['trades']  # route_trade_by_source
assert(f"{results[results.uid=='0']['amt1'].values[0]:.0f}" == "2519667817")
results

uid  id subid                                       note   aggr   exec  \
0   0.0   0     0                                   route #0  False  False   
0   0.1   0     1                                   route #2  False  False   
0   0.2   0     2                                   route #4  False  False   
0   0.3   0     3                                   route #6  False  False   
0   0.4   0     4                                   route #8  False  False   
0   0.5   0     5                                  route #18  False  False   
0   0.6   0     6                                  route #20  False  False   
0   0.7   0     7                                  route #14  False  False   
0   0.8   0     8                                  route #16  False  False   
0   0.9   0     9                                  route #12  False  False   
0  0.10   0    10                                  route #10  False  False   
0     0   0     A  AMM sells 2519667817USDC buys 10000000ETH   True  False   

  limitfail          amt1  tkn1          amt2 tkn2     pair  \
0      None  2.548147e+08  USDC  9.934316e+05  ETH  ETHUSDC   
0      None  2.538271e+08  USDC  9.934543e+05  ETH  ETHUSDC   
0      None  2.528394e+08  USDC  9.934771e+05  ETH  ETHUSDC   
0      None  2.518518e+08  USDC  9.935000e+05  ETH  ETHUSDC   
0      None  2.488888e+08  USDC  9.935701e+05  ETH  ETHUSDC   
0      None  2.498765e+08  USDC  9.935465e+05  ETH  ETHUSDC   
0      None  2.508641e+08  USDC  9.935232e+05  ETH  ETHUSDC   
0      None  2.469135e+08  USDC  9.936177e+05  ETH  ETHUSDC   
0      None  2.479012e+08  USDC  9.935938e+05  ETH  ETHUSDC   
0      None  2.459258e+08  USDC  9.936418e+05  ETH  ETHUSDC   
0      None  1.596493e+07  USDC  6.464387e+04  ETH  ETHUSDC   
0      None  2.519668e+09  USDC  1.000000e+07  ETH  ETHUSDC   

                                   routeix  nroutes       price        p_unit  
0                                        0        1  251.966782  USDC per ETH  
0                                        2        1  251.966782  USDC per ETH  
0                                        4        1  251.966782  USDC per ETH  
0                                        6        1  251.966782  USDC per ETH  
0                                        8        1  251.966782  USDC per ETH  
0                                       18        1  251.966782  USDC per ETH  
0                                       20        1  251.966782  USDC per ETH  
0                                       14        1  251.966782  USDC per ETH  
0                                       16        1  251.966782  USDC per ETH  
0                                       12        1  251.966782  USDC per ETH  
0                                       10        1  251.966782  USDC per ETH  
0  [0, 2, 4, 6, 8, 18, 20, 14, 16, 12, 10]       11  251.966782  USDC per ETH

We can observe the price is 251.996 USDC per ETH.

For trade_by_source, when we increase the trade amount we push further down the curve which means that effective price should get worse.

In [7]:
results = Sim.amm_buys('ETH',10500000, execute=False)['trades']  # route_trade_by_source
assert(f"{results[results.uid=='1']['amt1'].values[0]:.0f}" == "2643009195")
results

uid  id subid                                       note   aggr   exec  \
0   1.0   1     0                                   route #0  False  False   
0   1.1   1     1                                   route #2  False  False   
0   1.2   1     2                                   route #4  False  False   
0   1.3   1     3                                   route #6  False  False   
0   1.4   1     4                                   route #8  False  False   
0   1.5   1     5                                  route #18  False  False   
0   1.6   1     6                                  route #20  False  False   
0   1.7   1     7                                  route #14  False  False   
0   1.8   1     8                                  route #16  False  False   
0   1.9   1     9                                  route #12  False  False   
0  1.10   1    10                                  route #10  False  False   
0     1   1     A  AMM sells 2643009195USDC buys 10500000ETH   True  False   

  limitfail          amt1  tkn1          amt2 tkn2     pair  \
0      None  2.548147e+08  USDC  9.934316e+05  ETH  ETHUSDC   
0      None  2.538271e+08  USDC  9.934543e+05  ETH  ETHUSDC   
0      None  2.528394e+08  USDC  9.934771e+05  ETH  ETHUSDC   
0      None  2.518518e+08  USDC  9.935000e+05  ETH  ETHUSDC   
0      None  2.488888e+08  USDC  9.935701e+05  ETH  ETHUSDC   
0      None  2.498765e+08  USDC  9.935465e+05  ETH  ETHUSDC   
0      None  2.508641e+08  USDC  9.935232e+05  ETH  ETHUSDC   
0      None  2.469135e+08  USDC  9.936177e+05  ETH  ETHUSDC   
0      None  2.479012e+08  USDC  9.935938e+05  ETH  ETHUSDC   
0      None  2.459258e+08  USDC  9.936418e+05  ETH  ETHUSDC   
0      None  1.393063e+08  USDC  5.646439e+05  ETH  ETHUSDC   
0      None  2.643009e+09  USDC  1.050000e+07  ETH  ETHUSDC   

                                   routeix  nroutes       price        p_unit  
0                                        0        1  251.715161  USDC per ETH  
0                                        2        1  251.715161  USDC per ETH  
0                                        4        1  251.715161  USDC per ETH  
0                                        6        1  251.715161  USDC per ETH  
0                                        8        1  251.715161  USDC per ETH  
0                                       18        1  251.715161  USDC per ETH  
0                                       20        1  251.715161  USDC per ETH  
0                                       14        1  251.715161  USDC per ETH  
0                                       16        1  251.715161  USDC per ETH  
0                                       12        1  251.715161  USDC per ETH  
0                                       10        1  251.715161  USDC per ETH  
0  [0, 2, 4, 6, 8, 18, 20, 14, 16, 12, 10]       11  251.715161  USDC per ETH

And indeed for the exact algo, the price gets lower (251.996 vs 251.715), i.e. for every unit of ETH spent I get less units of USDC in return

We can then do a similar thing for the fast router

In [8]:
results = FastSim.amm_buys('ETH',10000000, execute=False)['trades']  # route_trade_by_source
assert(f"{results[results.uid=='0']['amt1'].values[0]:.0f}" == "2519667817")
results

uid  id subid                                       note   aggr   exec  \
0   0.0   0     0                                   route #0  False  False   
0   0.1   0     1                                   route #2  False  False   
0   0.2   0     2                                   route #4  False  False   
0   0.3   0     3                                   route #6  False  False   
0   0.4   0     4                                  route #20  False  False   
0   0.5   0     5                                  route #18  False  False   
0   0.6   0     6                                   route #8  False  False   
0   0.7   0     7                                  route #16  False  False   
0   0.8   0     8                                  route #14  False  False   
0   0.9   0     9                                  route #12  False  False   
0  0.10   0    10                                  route #10  False  False   
0     0   0     A  AMM sells 2519667817USDC buys 10000000ETH   True  False   

  limitfail          amt1  tkn1          amt2 tkn2     pair  \
0      None  2.548147e+08  USDC  9.934316e+05  ETH  ETHUSDC   
0      None  2.538271e+08  USDC  9.934543e+05  ETH  ETHUSDC   
0      None  2.528394e+08  USDC  9.934771e+05  ETH  ETHUSDC   
0      None  2.518518e+08  USDC  9.935000e+05  ETH  ETHUSDC   
0      None  2.508641e+08  USDC  9.935232e+05  ETH  ETHUSDC   
0      None  2.498765e+08  USDC  9.935465e+05  ETH  ETHUSDC   
0      None  2.488888e+08  USDC  9.935701e+05  ETH  ETHUSDC   
0      None  2.479012e+08  USDC  9.935938e+05  ETH  ETHUSDC   
0      None  2.469135e+08  USDC  9.936177e+05  ETH  ETHUSDC   
0      None  2.459258e+08  USDC  9.936418e+05  ETH  ETHUSDC   
0      None  1.596493e+07  USDC  6.464387e+04  ETH  ETHUSDC   
0      None  2.519668e+09  USDC  1.000000e+07  ETH  ETHUSDC   

                                   routeix  nroutes       price        p_unit  
0                                        0        1  251.966782  USDC per ETH  
0                                        2        1  251.966782  USDC per ETH  
0                                        4        1  251.966782  USDC per ETH  
0                                        6        1  251.966782  USDC per ETH  
0                                       20        1  251.966782  USDC per ETH  
0                                       18        1  251.966782  USDC per ETH  
0                                        8        1  251.966782  USDC per ETH  
0                                       16        1  251.966782  USDC per ETH  
0                                       14        1  251.966782  USDC per ETH  
0                                       12        1  251.966782  USDC per ETH  
0                                       10        1  251.966782  USDC per ETH  
0  [0, 2, 4, 6, 20, 18, 8, 16, 14, 12, 10]       11  251.966782  USDC per ETH

We see that the orders are filled the same and the corresponding price is as we saw before.

We can then trade the higher amount and again, expect to see the effectice rate decrease.

In [9]:
results = FastSim.amm_buys('ETH',10500000, execute=False)['trades']  # route_trade_by_source
assert(f"{results[results.uid=='1']['amt1'].values[0]:.0f}" == "2643009195")
results

uid  id subid                                       note   aggr   exec  \
0   1.0   1     0                                   route #0  False  False   
0   1.1   1     1                                   route #2  False  False   
0   1.2   1     2                                   route #4  False  False   
0   1.3   1     3                                   route #6  False  False   
0   1.4   1     4                                  route #20  False  False   
0   1.5   1     5                                  route #18  False  False   
0   1.6   1     6                                   route #8  False  False   
0   1.7   1     7                                  route #16  False  False   
0   1.8   1     8                                  route #14  False  False   
0   1.9   1     9                                  route #12  False  False   
0  1.10   1    10                                  route #10  False  False   
0     1   1     A  AMM sells 2643009195USDC buys 10500000ETH   True  False   

  limitfail          amt1  tkn1          amt2 tkn2     pair  \
0      None  2.548147e+08  USDC  9.934316e+05  ETH  ETHUSDC   
0      None  2.538271e+08  USDC  9.934543e+05  ETH  ETHUSDC   
0      None  2.528394e+08  USDC  9.934771e+05  ETH  ETHUSDC   
0      None  2.518518e+08  USDC  9.935000e+05  ETH  ETHUSDC   
0      None  2.508641e+08  USDC  9.935232e+05  ETH  ETHUSDC   
0      None  2.498765e+08  USDC  9.935465e+05  ETH  ETHUSDC   
0      None  2.488888e+08  USDC  9.935701e+05  ETH  ETHUSDC   
0      None  2.479012e+08  USDC  9.935938e+05  ETH  ETHUSDC   
0      None  2.469135e+08  USDC  9.936177e+05  ETH  ETHUSDC   
0      None  2.459258e+08  USDC  9.936418e+05  ETH  ETHUSDC   
0      None  1.393063e+08  USDC  5.646439e+05  ETH  ETHUSDC   
0      None  2.643009e+09  USDC  1.050000e+07  ETH  ETHUSDC   

                                   routeix  nroutes       price        p_unit  
0                                        0        1  251.715161  USDC per ETH  
0                                        2        1  251.715161  USDC per ETH  
0                                        4        1  251.715161  USDC per ETH  
0                                        6        1  251.715161  USDC per ETH  
0                                       20        1  251.715161  USDC per ETH  
0                                       18        1  251.715161  USDC per ETH  
0                                        8        1  251.715161  USDC per ETH  
0                                       16        1  251.715161  USDC per ETH  
0                                       14        1  251.715161  USDC per ETH  
0                                       12        1  251.715161  USDC per ETH  
0                                       10        1  251.715161  USDC per ETH  
0  [0, 2, 4, 6, 20, 18, 8, 16, 14, 12, 10]       11  251.715161  USDC per ETH

Again, we see that the price has, as before, decreased appropriately.

### Route by Target

#### AMM sells USDC

We can now test the route by target.

First we do the counter trade - selling the output amount of USDC from the initial ETH trade

In [10]:
results = Sim.amm_sells('USDC',2519667817, execute=False)['trades']  # route_trade_by_target
assert(f"{results[results.uid=='2']['amt2'].values[0]:.0f}" == "10000000")
results

uid  id subid                                       note   aggr   exec  \
0   2.0   2     0                                   route #0  False  False   
0   2.1   2     1                                   route #2  False  False   
0   2.2   2     2                                   route #4  False  False   
0   2.3   2     3                                   route #6  False  False   
0   2.4   2     4                                  route #18  False  False   
0   2.5   2     5                                  route #20  False  False   
0   2.6   2     6                                   route #8  False  False   
0   2.7   2     7                                  route #16  False  False   
0   2.8   2     8                                  route #14  False  False   
0   2.9   2     9                                  route #12  False  False   
0  2.10   2    10                                  route #10  False  False   
0     2   2     A  AMM sells 2519667817USDC buys 10000000ETH   True  False   

  limitfail          amt1  tkn1          amt2 tkn2     pair  \
0      None  2.548147e+08  USDC  9.934316e+05  ETH  ETHUSDC   
0      None  2.538271e+08  USDC  9.934543e+05  ETH  ETHUSDC   
0      None  2.528394e+08  USDC  9.934771e+05  ETH  ETHUSDC   
0      None  2.518518e+08  USDC  9.935000e+05  ETH  ETHUSDC   
0      None  2.498765e+08  USDC  9.935465e+05  ETH  ETHUSDC   
0      None  2.508641e+08  USDC  9.935232e+05  ETH  ETHUSDC   
0      None  2.488888e+08  USDC  9.935701e+05  ETH  ETHUSDC   
0      None  2.479012e+08  USDC  9.935938e+05  ETH  ETHUSDC   
0      None  2.469135e+08  USDC  9.936177e+05  ETH  ETHUSDC   
0      None  2.459258e+08  USDC  9.936418e+05  ETH  ETHUSDC   
0      None  1.596493e+07  USDC  6.464387e+04  ETH  ETHUSDC   
0      None  2.519668e+09  USDC  1.000000e+07  ETH  ETHUSDC   

                                   routeix  nroutes       price        p_unit  
0                                        0        1  251.966782  USDC per ETH  
0                                        2        1  251.966782  USDC per ETH  
0                                        4        1  251.966782  USDC per ETH  
0                                        6        1  251.966782  USDC per ETH  
0                                       18        1  251.966782  USDC per ETH  
0                                       20        1  251.966782  USDC per ETH  
0                                        8        1  251.966782  USDC per ETH  
0                                       16        1  251.966782  USDC per ETH  
0                                       14        1  251.966782  USDC per ETH  
0                                       12        1  251.966782  USDC per ETH  
0                                       10        1  251.966782  USDC per ETH  
0  [0, 2, 4, 6, 18, 20, 8, 16, 14, 12, 10]       11  251.966782  USDC per ETH

And we see that the return amount is 10000000 we saw before - but importantly so is the price

Now we expect that if we take less off the curve then the rate gets better

In [11]:
results = Sim.amm_sells('USDC',1500000000, execute=False)['trades']  # route_trade_by_target
assert(f"{results[results.uid=='3']['amt2'].values[0]:.0f}" == "5904869")
results

uid  id subid                                      note   aggr   exec  \
0  3.0   3     0                                  route #0  False  False   
0  3.1   3     1                                  route #2  False  False   
0  3.2   3     2                                  route #4  False  False   
0  3.3   3     3                                  route #6  False  False   
0  3.4   3     4                                 route #20  False  False   
0  3.5   3     5                                 route #18  False  False   
0    3   3     A  AMM sells 1500000000USDC buys 5904869ETH   True  False   

  limitfail          amt1  tkn1          amt2 tkn2     pair  \
0      None  2.548147e+08  USDC  9.934316e+05  ETH  ETHUSDC   
0      None  2.538271e+08  USDC  9.934543e+05  ETH  ETHUSDC   
0      None  2.528394e+08  USDC  9.934771e+05  ETH  ETHUSDC   
0      None  2.518518e+08  USDC  9.935000e+05  ETH  ETHUSDC   
0      None  2.508641e+08  USDC  9.935232e+05  ETH  ETHUSDC   
0      None  2.358029e+08  USDC  9.374828e+05  ETH  ETHUSDC   
0      None  1.500000e+09  USDC  5.904869e+06  ETH  ETHUSDC   

                routeix  nroutes       price        p_unit  
0                     0        1  254.027651  USDC per ETH  
0                     2        1  254.027651  USDC per ETH  
0                     4        1  254.027651  USDC per ETH  
0                     6        1  254.027651  USDC per ETH  
0                    20        1  254.027651  USDC per ETH  
0                    18        1  254.027651  USDC per ETH  
0  [0, 2, 4, 6, 20, 18]        6  254.027651  USDC per ETH

Indeed we see that the price is better for a smaller amount is less (254.027 vs 251.996)

Now we do the same for the Fast router. Trade in the previous output amount

In [12]:
results = FastSim.amm_sells('USDC',2519667817, execute=False)['trades']  # route_trade_by_target
assert(f"{results[results.uid=='2']['amt2'].values[0]:.0f}" == "10000000")
results

uid  id subid                                       note   aggr   exec  \
0   2.0   2     0                                   route #0  False  False   
0   2.1   2     1                                   route #2  False  False   
0   2.2   2     2                                   route #4  False  False   
0   2.3   2     3                                   route #6  False  False   
0   2.4   2     4                                  route #20  False  False   
0   2.5   2     5                                  route #18  False  False   
0   2.6   2     6                                   route #8  False  False   
0   2.7   2     7                                  route #16  False  False   
0   2.8   2     8                                  route #14  False  False   
0   2.9   2     9                                  route #12  False  False   
0  2.10   2    10                                  route #10  False  False   
0     2   2     A  AMM sells 2519667817USDC buys 10000000ETH   True  False   

  limitfail          amt1  tkn1          amt2 tkn2     pair  \
0      None  2.548147e+08  USDC  9.934316e+05  ETH  ETHUSDC   
0      None  2.538271e+08  USDC  9.934543e+05  ETH  ETHUSDC   
0      None  2.528394e+08  USDC  9.934771e+05  ETH  ETHUSDC   
0      None  2.518518e+08  USDC  9.935000e+05  ETH  ETHUSDC   
0      None  2.508641e+08  USDC  9.935232e+05  ETH  ETHUSDC   
0      None  2.498765e+08  USDC  9.935465e+05  ETH  ETHUSDC   
0      None  2.488888e+08  USDC  9.935701e+05  ETH  ETHUSDC   
0      None  2.479012e+08  USDC  9.935938e+05  ETH  ETHUSDC   
0      None  2.469135e+08  USDC  9.936177e+05  ETH  ETHUSDC   
0      None  2.459258e+08  USDC  9.936418e+05  ETH  ETHUSDC   
0      None  1.596493e+07  USDC  6.464387e+04  ETH  ETHUSDC   
0      None  2.519668e+09  USDC  1.000000e+07  ETH  ETHUSDC   

                                   routeix  nroutes       price        p_unit  
0                                        0        1  251.966782  USDC per ETH  
0                                        2        1  251.966782  USDC per ETH  
0                                        4        1  251.966782  USDC per ETH  
0                                        6        1  251.966782  USDC per ETH  
0                                       20        1  251.966782  USDC per ETH  
0                                       18        1  251.966782  USDC per ETH  
0                                        8        1  251.966782  USDC per ETH  
0                                       16        1  251.966782  USDC per ETH  
0                                       14        1  251.966782  USDC per ETH  
0                                       12        1  251.966782  USDC per ETH  
0                                       10        1  251.966782  USDC per ETH  
0  [0, 2, 4, 6, 20, 18, 8, 16, 14, 12, 10]       11  251.966782  USDC per ETH

And importantly we see the same price to the exact algo

The test now is to see the price get better as the request amount is lowered

In [13]:
results = FastSim.amm_sells('USDC',1500000000, execute=False)['trades']  # route_trade_by_target
assert(f"{results[results.uid=='3']['amt2'].values[0]:.0f}" == "5904869")
results

uid  id subid                                      note   aggr   exec  \
0  3.0   3     0                                  route #0  False  False   
0  3.1   3     1                                  route #2  False  False   
0  3.2   3     2                                  route #4  False  False   
0  3.3   3     3                                  route #6  False  False   
0  3.4   3     4                                 route #20  False  False   
0  3.5   3     5                                 route #18  False  False   
0    3   3     A  AMM sells 1500000000USDC buys 5904869ETH   True  False   

  limitfail          amt1  tkn1          amt2 tkn2     pair  \
0      None  2.548147e+08  USDC  9.934316e+05  ETH  ETHUSDC   
0      None  2.538271e+08  USDC  9.934543e+05  ETH  ETHUSDC   
0      None  2.528394e+08  USDC  9.934771e+05  ETH  ETHUSDC   
0      None  2.518518e+08  USDC  9.935000e+05  ETH  ETHUSDC   
0      None  2.508641e+08  USDC  9.935232e+05  ETH  ETHUSDC   
0      None  2.358029e+08  USDC  9.374828e+05  ETH  ETHUSDC   
0      None  1.500000e+09  USDC  5.904869e+06  ETH  ETHUSDC   

                routeix  nroutes       price        p_unit  
0                     0        1  254.027651  USDC per ETH  
0                     2        1  254.027651  USDC per ETH  
0                     4        1  254.027651  USDC per ETH  
0                     6        1  254.027651  USDC per ETH  
0                    20        1  254.027651  USDC per ETH  
0                    18        1  254.027651  USDC per ETH  
0  [0, 2, 4, 6, 20, 18]        6  254.027651  USDC per ETH

And indeed the rate gets better